File: python/src/sec_edgar_toolkit/core/xbrl/linkbases.py
"""
Parsers for XBRL presentation and label linkbases.

The presentation linkbase (``<base>_pre.xml``) defines, per statement
role, which concepts appear and in what order and hierarchy. The label
linkbase (``<base>_lab.xml``) carries the human-readable labels,
including the preferred variants (terse, total, negated) that the
presentation arcs select.

Standard library only. No I/O happens in this module.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple
from xml.etree import ElementTree as ET

LINK_NS = "http://www.xbrl.org/2003/linkbase"
XLINK_NS = "http://www.w3.org/1999/xlink"

STANDARD_LABEL = "http://www.xbrl.org/2003/role/label"


def _concept_from_href(href: str) -> str:
    """``...us-gaap-2025.xsd#us-gaap_Revenues`` -> ``us-gaap:Revenues``."""
    fragment = href.rsplit("#", 1)[-1]
    return fragment.replace("_", ":", 1)


@dataclass
class PresentationNode:
    """One concept's position within a statement role."""

    concept: str
    order: float = 0.0
    depth: int = 0
    preferred_label: Optional[str] = None
    children: List["PresentationNode"] = field(default_factory=list)


class PresentationLinkbase:
    """Per-role concept ordering parsed from ``<base>_pre.xml``."""

    def __init__(self, roles: Dict[str, List[PresentationNode]]) -> None:
        #: role URI -> root nodes in presentation order
        self.roles = roles

    @classmethod
    def parse(cls, raw: "bytes | str") -> "PresentationLinkbase":
        if isinstance(raw, str):
            raw = raw.encode("utf-8", errors="ignore")
        root = ET.fromstring(raw)

        roles: Dict[str, List[PresentationNode]] = {}
        for link in root.findall(f"{{{LINK_NS}}}presentationLink"):
            role = link.get(f"{{{XLINK_NS}}}role", "")
            if not role:
                continue
            roles[role] = cls._parse_link(link)
        return cls(roles)

    @classmethod
    def _parse_link(cls, link: ET.Element) -> List[PresentationNode]:
        locators: Dict[str, str] = {}
        for loc in link.findall(f"{{{LINK_NS}}}loc"):
            label = loc.get(f"{{{XLINK_NS}}}label", "")
            href = loc.get(f"{{{XLINK_NS}}}href", "")
            if label and href:
                locators[label] = _concept_from_href(href)

        #: parent locator label -> [(order, child locator label, preferred)]
        arcs: Dict[str, List[Tuple[float, str, Optional[str]]]] = {}
        children_labels: set = set()
        for arc in link.findall(f"{{{LINK_NS}}}presentationArc"):
            parent = arc.get(f"{{{XLINK_NS}}}from", "")
            child = arc.get(f"{{{XLINK_NS}}}to", "")
            if not parent or not child:
                continue
            try:
                order = float(arc.get("order", "0"))
            except ValueError:
                order = 0.0
            arcs.setdefault(parent, []).append(
                (order, child, arc.get("preferredLabel"))
            )
            children_labels.add(child)

        def build(label: str, depth: int, preferred: Optional[str]) -> PresentationNode:
            node = PresentationNode(
                concept=locators.get(label, label),
                depth=depth,
                preferred_label=preferred,
            )
            for order, child_label, child_preferred in sorted(
                arcs.get(label, []), key=lambda entry: entry[0]
            ):
                child = build(child_label, depth + 1, child_preferred)
                child.order = order
                node.children.append(child)
            return node

        roots = [
            build(label, 0, None)
            for label in locators
            if label not in children_labels and label in arcs
        ]
        return roots

    def ordered_concepts(self, role: str) -> List[PresentationNode]:
        """Depth-first, presentation-ordered nodes for one role."""
        flattened: List[PresentationNode] = []

        def walk(node: PresentationNode) -> None:
            flattened.append(node)
            for child in node.children:
                walk(child)

        for node in self.roles.get(role, []):
            walk(node)
        return flattened

    def find_role(self, role_or_suffix: str) -> Optional[str]:
        """Resolve a full role URI or a suffix of one."""
        if role_or_suffix in self.roles:
            return role_or_suffix
        needle = role_or_suffix.lower()
        for role in self.roles:
            if role.lower().endswith(needle):
                return role
        return None


class LabelLinkbase:
    """Concept labels parsed from ``<base>_lab.xml``."""

    def __init__(self, labels: Dict[str, Dict[str, str]]) -> None:
        #: concept qname -> {label role URI: text}
        self.labels = labels

    @classmethod
    def parse(cls, raw: "bytes | str") -> "LabelLinkbase":
        if isinstance(raw, str):
            raw = raw.encode("utf-8", errors="ignore")
        root = ET.fromstring(raw)

        labels: Dict[str, Dict[str, str]] = {}
        for link in root.findall(f"{{{LINK_NS}}}labelLink"):
            locators: Dict[str, str] = {}
            for loc in link.findall(f"{{{LINK_NS}}}loc"):
                label = loc.get(f"{{{XLINK_NS}}}label", "")
                href = loc.get(f"{{{XLINK_NS}}}href", "")
                if label and href:
                    locators[label] = _concept_from_href(href)

            resources: Dict[str, Dict[str, str]] = {}
            for resource in link.findall(f"{{{LINK_NS}}}label"):
                resource_label = resource.get(f"{{{XLINK_NS}}}label", "")
                role = resource.get(f"{{{XLINK_NS}}}role", STANDARD_LABEL)
                if resource_label and resource.text:
                    resources.setdefault(resource_label, {})[role] = resource.text

            for arc in link.findall(f"{{{LINK_NS}}}labelArc"):
                concept = locators.get(arc.get(f"{{{XLINK_NS}}}from", ""))
                label_texts = resources.get(arc.get(f"{{{XLINK_NS}}}to", ""))
                if concept and label_texts:
                    labels.setdefault(concept, {}).update(label_texts)

        return cls(labels)

    def label_for(self, concept: str, preferred: Optional[str] = None) -> Optional[str]:
        """The label for a concept, honoring a preferred label role."""
        concept_labels = self.labels.get(concept)
        if not concept_labels:
            return None
        if preferred and preferred in concept_labels:
            return concept_labels[preferred]
        if STANDARD_LABEL in concept_labels:
            return concept_labels[STANDARD_LABEL]
        return next(iter(concept_labels.values()))

File: typescript/src/core/xbrl/linkbases.ts
/**
 * Parsers for XBRL presentation and label linkbases.
 *
 * The presentation linkbase (`<base>_pre.xml`) defines, per statement
 * role, which concepts appear and in what order and hierarchy. The label
 * linkbase (`<base>_lab.xml`) carries the human-readable labels,
 * including the preferred variants that the presentation arcs select.
 */

import { XMLParser } from "fast-xml-parser";

export const STANDARD_LABEL = "http://www.xbrl.org/2003/role/label";

export interface PresentationNode {
  concept: string;
  order: number;
  depth: number;
  preferredLabel: string | null;
  children: PresentationNode[];
}

function asArray<T>(value: T | T[] | undefined): T[] {
  if (value === undefined || value === null) {
    return [];
  }
  return Array.isArray(value) ? value : [value];
}

function localName(tag: string): string {
  return tag.includes(":") ? tag.slice(tag.indexOf(":") + 1) : tag;
}

function findKey(obj: Record<string, any>, local: string): any {
  for (const [key, value] of Object.entries(obj)) {
    if (localName(key) === local) {
      return value;
    }
  }
  return undefined;
}

function conceptFromHref(href: string): string {
  const fragment = href.split("#").pop() || "";
  const underscore = fragment.indexOf("_");
  if (underscore < 0) {
    return fragment;
  }
  return `${fragment.slice(0, underscore)}:${fragment.slice(underscore + 1)}`;
}

function parseXml(raw: string): Record<string, any> {
  const parser = new XMLParser({
    ignoreAttributes: false,
    parseTagValue: false,
    trimValues: true,
  });
  const document = parser.parse(raw);
  const rootKey = Object.keys(document).find(
    (k) => localName(k) === "linkbase",
  );
  if (!rootKey) {
    throw new Error("Not an XBRL linkbase");
  }
  return document[rootKey];
}

export class PresentationLinkbase {
  /** role URI -> root nodes in presentation order */
  public readonly roles: Map<string, PresentationNode[]>;

  private constructor(roles: Map<string, PresentationNode[]>) {
    this.roles = roles;
  }

  static parse(raw: string): PresentationLinkbase {
    const root = parseXml(raw);
    const roles = new Map<string, PresentationNode[]>();

    for (const link of asArray(findKey(root, "presentationLink"))) {
      const role = String(link["@_xlink:role"] || "");
      if (!role) {
        continue;
      }
      roles.set(role, PresentationLinkbase.parseLink(link));
    }
    return new PresentationLinkbase(roles);
  }

  private static parseLink(link: Record<string, any>): PresentationNode[] {
    const locators = new Map<string, string>();
    for (const loc of asArray(findKey(link, "loc"))) {
      const label = String(loc["@_xlink:label"] || "");
      const href = String(loc["@_xlink:href"] || "");
      if (label && href) {
        locators.set(label, conceptFromHref(href));
      }
    }

    const arcs = new Map<
      string,
      Array<{ order: number; child: string; preferred: string | null }>
    >();
    const childLabels = new Set<string>();
    for (const arc of asArray(findKey(link, "presentationArc"))) {
      const parent = String(arc["@_xlink:from"] || "");
      const child = String(arc["@_xlink:to"] || "");
      if (!parent || !child) {
        continue;
      }
      const order = Number(arc["@_order"] || 0) || 0;
      const list = arcs.get(parent) || [];
      list.push({
        order,
        child,
        preferred: arc["@_preferredLabel"]
          ? String(arc["@_preferredLabel"])
          : null,
      });
      arcs.set(parent, list);
      childLabels.add(child);
    }

    const build = (
      label: string,
      depth: number,
      preferred: string | null,
    ): PresentationNode => {
      const node: PresentationNode = {
        concept: locators.get(label) || label,
        order: 0,
        depth,
        preferredLabel: preferred,
        children: [],
      };
      const children = (arcs.get(label) || [])
        .slice()
        .sort((a, b) => a.order - b.order);
      for (const entry of children) {
        const child = build(entry.child, depth + 1, entry.preferred);
        child.order = entry.order;
        node.children.push(child);
      }
      return node;
    };

    const roots: PresentationNode[] = [];
    for (const label of locators.keys()) {
      if (!childLabels.has(label) && arcs.has(label)) {
        roots.push(build(label, 0, null));
      }
    }
    return roots;
  }

  orderedConcepts(role: string): PresentationNode[] {
    const flattened: PresentationNode[] = [];
    const walk = (node: PresentationNode): void => {
      flattened.push(node);
      for (const child of node.children) {
        walk(child);
      }
    };
    for (const node of this.roles.get(role) || []) {
      walk(node);
    }
    return flattened;
  }

  findRole(roleOrSuffix: string): string | null {
    if (this.roles.has(roleOrSuffix)) {
      return roleOrSuffix;
    }
    const needle = roleOrSuffix.toLowerCase();
    for (const role of this.roles.keys()) {
      if (role.toLowerCase().endsWith(needle)) {
        return role;
      }
    }
    return null;
  }
}

export class LabelLinkbase {
  /** concept qname -> {label role URI: text} */
  public readonly labels: Map<string, Record<string, string>>;

  private constructor(labels: Map<string, Record<string, string>>) {
    this.labels = labels;
  }

  static parse(raw: string): LabelLinkbase {
    const root = parseXml(raw);
    const labels = new Map<string, Record<string, string>>();

    for (const link of asArray(findKey(root, "labelLink"))) {
      const locators = new Map<string, string>();
      for (const loc of asArray(findKey(link, "loc"))) {
        const label = String(loc["@_xlink:label"] || "");
        const href = String(loc["@_xlink:href"] || "");
        if (label && href) {
          locators.set(label, conceptFromHref(href));
        }
      }

      const resources = new Map<string, Record<string, string>>();
      for (const resource of asArray(findKey(link, "label"))) {
        const resourceLabel = String(resource["@_xlink:label"] || "");
        const role = String(resource["@_xlink:role"] || STANDARD_LABEL);
        const value =
          typeof resource === "object"
            ? String(resource["#text"] ?? "")
            : String(resource);
        if (resourceLabel && value) {
          const entry = resources.get(resourceLabel) || {};
          entry[role] = value;
          resources.set(resourceLabel, entry);
        }
      }

      for (const arc of asArray(findKey(link, "labelArc"))) {
        const concept = locators.get(String(arc["@_xlink:from"] || ""));
        const labelTexts = resources.get(String(arc["@_xlink:to"] || ""));
        if (concept && labelTexts) {
          const entry = labels.get(concept) || {};
          Object.assign(entry, labelTexts);
          labels.set(concept, entry);
        }
      }
    }

    return new LabelLinkbase(labels);
  }

  labelFor(concept: string, preferred?: string | null): string | null {
    const conceptLabels = this.labels.get(concept);
    if (!conceptLabels) {
      return null;
    }
    if (preferred && conceptLabels[preferred]) {
      return conceptLabels[preferred];
    }
    if (conceptLabels[STANDARD_LABEL]) {
      return conceptLabels[STANDARD_LABEL];
    }
    const first = Object.values(conceptLabels)[0];
    return first ?? null;
  }
}
